WIP: add advanced features of paths.

togawa28 opened the pull request:
The results of exploring plots for features we have now (distance, speeds,...) were somewhat disappointing; probably we need to add more features which should indicate significant and interesting differences over different strains. This pull request will add the features such as: 
1. area of the rectangle spanned by path
2. radius and center angles in path
3. area covered by path
4. distance between the start and end points in path
For details, see #122.

For the initial commit, I only added main functions to generate those features. I'll add test functions later. Please review @aksam-ahmad @alannaiverson @jcpalumbo3 @jJasonWang @sicihuang

alannaiverson commented on mousestyles/path_diversity/path_features_advanced.py:
```diff
@@ -0,0 +1,287 @@
+
+    Returns
+    -------
+    center point of the path : dictinoary
```
just a simple typo (should be dictionary) but it might be important since it's a type

togawa28 replied:
Thanks. I'll fix it in the next revision.

alannaiverson commented on mousestyles/path_diversity/path_features_advanced.py:
```diff
@@ -0,0 +1,287 @@
+    areas = [ v1 * v2 * np.sin(theta) / 2 for v1,v2,theta in zipped]
+    return(sum(areas))
+
+def compute_start_end_distance(path_obj):
```
Doesn't the function @jcpalumbo3 and @sicihuang compute this quantity if desired?

togawa28 replied:
This function does not calculate the total length of the path nor the distances between adjacent 2 points, but does calculate the length of the straight line which directly connects the starting and ending points in a path. In other words, we ignore the intermediate points and only focus on the 2 points, the 1st and the last. I think this can't be done by `get_dist_and_speed` made by @jcpalumbo3 and @sicihuang.

alannaiverson commented:
@togawa28 What if we only calculate two of the features, like area of rectangle and radius, but only have one function that returns both? I think that could work and testing would be significantly less? What do y'all think, @jcpalumbo3 @jJasonWang @aksam-ahmad @sicihuang ?

togawa28 commented:
Interesting idea @alannaiverson, we could combine some functions, but why do you think testing would be less? Regardless of using one or several, we need to test the inputs/outputs of area of rectangle and radius anyway?

alannaiverson commented:
True, @togawa28 but I was thinking if we just computed the edge points and center within the function, those two at least wouldn't need separate testing would they? Or am I misunderstanding?

togawa28 commented:
No, you correctly understand what I'm saying. Well, even if we include those two within the function, we still should test those so that we confirm they work correctly maybe?

alannaiverson commented:
I think testing the final output would serve as a check for those, without formal tests, but I could be wrong. I was thinking something like this (not sure why markdown is making the first few lines weird...): 
`
def compute_advanced(path_obj):
    if not isinstance(path_obj, pd.core.frame.DataFrame):
        raise TypeError("path_obj must be pandas DataFrame")

    if not set(path_obj.keys()).issuperset(['x', 'y']):
        raise ValueError("the keys of path_obj must contain 'x', 'y'")

    if len(path_obj) <= 1:
        raise ValueError("path_obj must contain at least 2 rows")

    # Computes edge points like find_edge_points originally did compute_area_rectangle did
    edge_points = {'xmin': np.min(path_obj.x), 'xmax': np.max(path_obj.x),
                   'ymin': np.min(path_obj.y), 'ymax': np.max(path_obj.y)}

    # Computes area of rectangle
    area = (edge_points['xmax']-edge_points['xmin']) * (edge_points['ymax']-edge_points['ymin'])
    
    # Computes center like find_center originally did
    center = {'x': (edge_points['xmin'] + edge_points['xmax'])/2,
              'y': (edge_points['ymin'] + edge_points['ymax'])/2}
    
    indices = path_obj.index[:-1]
    vectors = [path_obj.loc[i, 'x':'y'] -
               [center['x'], center['y']] for i in indices]
    center_angles = [path_features.angle_between(list(v1), list(v2)) for
                     v1, v2 in zip(vectors[1:], vectors[:len(vectors)])]
    radius = [np.linalg.norm(v) for v in vectors]
    
    return({'radius': radius, 'center_angles': center_angles, 'area': area})`

@togawa28 @jJasonWang @jcpalumbo3 @sicihuang @aksam-ahmad Thoughts on squashing some of the functions into one for time reasons?

togawa28 commented:
@alannaiverson thank you for your contribution, and now I get your point. It's much more efficient to squash those functions because it not only decreases the number of tedious testing functions but also  reduces doc strings to make the whole file easier to read and understand.
I'll follow your suggestion and start writing tests functions. Thanks again!

togawa28 commented:
BTW, for PEP8 styles issue, I got the following message.

________________________________ pyflakes-check ________________________________
/home/travis/build/berkeley-stat222/mousestyles/mousestyles/path_diversity/path_features_advanced.py:84: EOL while scanning string literal
        raise ValueError("the keys of edge_points must be 'xmax', 'xmin',

I'm not sure how to fix it. Any idea?

jJasonWang commented:
I am not sure what happens, but I will help you check that @togawa28.

alannaiverson commented:
@togawa28 I made an example of what the function and tests might look like, but I'm not sure how to add a file to this pull request so I'll just copy them here? Apologies for the messiness and these are just ideas! They can no doubt be improved:

`from __future__ import print_function, absolute_import, division
import numpy as np
import pandas as pd
from mousestyles.path_diversity.path_features import angle_between
from mousestyles.path_diversity import path_index


def compute_advanced(path_obj):
    r"""
    Return dictionary containing the radius,  area of the rectangle 
    spanned by the path, and absolute distance between start and end
    points of each path.

    Parameters
    ----------
    path_obj : pandas.DataFrame
        CX and CY must be contained.
        The length must be greater than 2.

    Returns
    -------
    radius : numpy float object
        each element is the distance between center point and
        each point in the path. The length equals to the length
        of the path_obj.
    
    area : numpy float object
        area of the rectangle spanned by the path.
        
    abs_distance : numpy float
        the distance between the start and end points in a path.

    Examples
    --------
    >>> movement = data.load_movement(1,2,1)
    >>> compute_advanced(movement)
    """
    if not isinstance(path_obj, pd.core.frame.DataFrame):
        raise TypeError("path_obj must be pandas DataFrame")

    if not set(path_obj.keys()).issuperset(['x', 'y']):
        raise ValueError("the keys of path_obj must contain 'x', 'y'")

    if len(path_obj) <= 1:
        raise ValueError("path_obj must contain at least 2 rows")

    # Computes edge points like find_edge_points originally did compute_area_rectangle did
    edge_points = {'xmin': np.min(path_obj.x), 'xmax': np.max(path_obj.x),
                   'ymin': np.min(path_obj.y), 'ymax': np.max(path_obj.y)}

    # Computes area of rectangle
    area = (edge_points['xmax']-edge_points['xmin']) * (edge_points['ymax']-edge_points['ymin'])
    
    # Computes center like find_center originally did
    center = {'x': (edge_points['xmin'] + edge_points['xmax'])/2,
              'y': (edge_points['ymin'] + edge_points['ymax'])/2}
    
    # Computes average radius
    indices = path_obj.index[:-1]
    vectors = [path_obj.loc[i, 'x':'y'] -
               [center['x'], center['y']] for i in indices]
    radius = [np.linalg.norm(v) for v in vectors]
    # but maybe we just want a single number for this? We could do average or sum?
    avg_radius = sum(radius)/len(radius)
    
    # Computes distance between start and end points of each path
    initial = path_obj.loc[path_obj.index[0], 'x':'y']
    end = path_obj.loc[path_obj.index[-1], 'x':'y']
    abs_distance = np.sqrt((end['x'] - initial['x'])**2 +
                           (end['y'] - initial['y'])**2)
    
    return({'radius': avg_radius, 'area': area, 'abs_distance': abs_distance})`

and tests: 
`from __future__ import (absolute_import, division, print_function,
                        unicode_literals)

import pytest

from mousestyles import data
from mousestyles.path_diversity import path_index
from mousestyles.path_diversity import filter_path


def test_compute_advanced_input():
    # Check if function raises the correct type of errors.
    path = pd.DataFrame({'t': [2, 1, -1], 'x': [5, 3, 8],
                         'y': [-3, 0, 0], 'isHB': [True, False, False]})
    # not pandas DataFrame
    like_path = [1, 2, 3]
    # not having keys 'x', 'y'
    like_path2 = pd.DataFrame(
        {'x': [5, -2, 1], 't': [-2, 3, 1], 'isHB': [True, True, False]})
    # length is less than 2
    like_path3 = pd.DataFrame({'t': [2], 'x': [5], 'y': [3], 'isHB': [True]})

    with pytest.raises(TypeError) as excinfo:
        path_features.compute_angles(like_path, True)
    assert excinfo.value.args[0] == "path_obj must be pandas DataFrame"

    with pytest.raises(ValueError) as excinfo:
        path_features.compute_angles(like_path2, True)
    assert excinfo.value.args[
        0] == "the keys of path_obj must contain 'x', 'y'"

    with pytest.raises(ValueError) as excinfo:
        path_features.compute_angles(like_path3, True)
    assert excinfo.value.args[0] == "path_obj must contain at least 2 rows"

    
def test_compute_advanced():
    path = pd.DataFrame({'t': [2, 4.5, 10.5], 'x': [0, 1, 1], 'y': [
                        0, 0, 1], 'isHB': [True, True, False]})
    assert path_features.compute_advanced(path['area']) == 1
    assert path_features.compute_advanced(path['abs_distance']) == 1.4142135623730951
    assert path_features.compute_advanced(path['radius']) == 0.70710678118654757`

jJasonWang commented on mousestyles/path_diversity/path_features_advanced.py:
```diff
@@ -0,0 +1,297 @@
+    if set(edge_points.keys()) != {'xmax', 'xmin', 'ymax', 'ymin'}:
+        raise ValueError("the keys of edge_points must be 'xmax', 'xmin',
+                         'ymax', and 'ymin'")
+
```
I tried to add a back slash here and the error message disappeared. @togawa28 
```
 raise ValueError("the keys of edge_points must be 'xmax', 'xmin', \
                            'ymax', and 'ymin'")
```

jcpalumbo3 replied:
We're going to have to be careful with that in the test function. It'll probably recognize that string as containing a ton of spaces. One way around this is to set `error = "the keys of edge_points must be 'xmax', 'xmin', 'ymax', and 'min'" ` (if it fits on one line) and then put `raise ValueError(error)`.

togawa28 replied:
I'll try it in the next revision. Thanks.

jcpalumbo3 replied:
Did you try putting it above the if statement? Or try `err` or another really short naming scheme? @togawa28

togawa28 commented:
I added new temporary files which are just copies of @alannaiverson's comment. We can discuss on those new files now.

togawa28 commented on mousestyles/path_diversity/path_features_advanced.py:
```diff
@@ -0,0 +1,297 @@
+        raise TypeError("edge_points must be a dictionary")
+
+    if set(edge_points.keys()) != {'xmax', 'xmin', 'ymax', 'ymin'}:
+        raise ValueError('the keys of edge_points must be "xmax", \
```
As pointed out @jJasonWang and @jcpalumbo3 we should take care of this line?

togawa28 commented on mousestyles/path_diversity/path_features_advanced.py:
```diff
@@ -0,0 +1,297 @@
+        raise TypeError("edge_points must be a dictionary")
+
+    if set(edge_points.keys()) != {'xmax', 'xmin', 'ymax', 'ymin'}:
+        raise ValueError("the keys of edge_points must be \
```
For self notes, same as above.

togawa28 commented on mousestyles/path_diversity/path_features_advanced_proposed.py:
```diff
@@ -0,0 +1,71 @@
+    r"""
+    Return dictionary containing the radius, area of the rectangle 
+    spanned by the path, and absolute distance between start and end
+    points of each path.
```
Why did you select those features @alannaiverson? Is it just an example or do you prefer those?

alannaiverson replied:
A little of both - just trying to think what would be easily interpretable and relatively easy to calculate. Can definitely be changed since this is just a starting point!

togawa28 replied:
Currently we are not really sure which features will show significant difference over the strains. So it may make sense to include all features (including center angles and area covered in path) here and then to include only useful features we find interesting in plotting?

togawa28 commented on mousestyles/path_diversity/path_features_advanced_proposed.py:
```diff
@@ -0,0 +1,71 @@
+    Examples
+    --------
+    >>> movement = data.load_movement(1,2,1)
+    >>> compute_advanced(movement)
```
Probably should be changed to `adv_features = compute_advanced(movement)`?

togawa28 commented on mousestyles/path_diversity/path_features_advanced_proposed.py:
```diff
@@ -0,0 +1,71 @@
+import numpy as np
+import pandas as pd
+from mousestyles.path_diversity.path_features import angle_between
+from mousestyles.path_diversity import path_index
```
We don't use `path_index` here so we can remove this line.

togawa28 commented on mousestyles/path_diversity/path_features_advanced_proposed.py:
```diff
@@ -0,0 +1,71 @@
+    vectors = [path_obj.loc[i, 'x':'y'] -
+               [center['x'], center['y']] for i in indices]
+    radius = [np.linalg.norm(v) for v in vectors]
+    # but maybe we just want a single number for this? We could do average or sum?
```
This is a good point and I'm not sure what is the best reduction technique in this situation. Just returning `radius` itself as a list would be useful so that we can later choose average or sum or others when plotting?

togawa28 commented on mousestyles/path_diversity/tests/test_path_features_advanced_proposed.py:
```diff
@@ -0,0 +1,39 @@
+    with pytest.raises(ValueError) as excinfo:
+        path_features.compute_angles(like_path3, True)
+    assert excinfo.value.args[0] == "path_obj must contain at least 2 rows"
+    def test_compute_advanced():
```
The contents should be indented. I'll fix it in the next

togawa28 commented on mousestyles/path_diversity/tests/test_path_features_advanced_proposed.py:
```diff
@@ -0,0 +1,39 @@
+    def test_compute_advanced():
+    path = pd.DataFrame({'t': [2, 4.5, 10.5], 'x': [0, 1, 1], 'y': [
+    0, 0, 1], 'isHB': [True, True, False]})
+    assert path_features.compute_advanced(path['area']) == 1
```
`path_features.compute_advanced(path)['area']`

togawa28 commented on mousestyles/path_diversity/tests/test_path_features_advanced_proposed.py:
```diff
@@ -0,0 +1,39 @@
+    path = pd.DataFrame({'t': [2, 4.5, 10.5], 'x': [0, 1, 1], 'y': [
+    0, 0, 1], 'isHB': [True, True, False]})
+    assert path_features.compute_advanced(path['area']) == 1
+    assert path_features.compute_advanced(path['abs_distance']) == 1.4142135623730951
```
Could we use `np.sqrt(2)` instead?

sicihuang commented on mousestyles/path_diversity/path_features_advanced.py:
```diff
@@ -0,0 +1,297 @@
+from __future__ import print_function, absolute_import, division
+import numpy as np
+import pandas as pd
+# from mousestyles.path_features import path_features
```
Just making sure, you will be using the line that's been commented out in the final version right? Also, since all function files are under the path_diversity folder, should it be `from mousestyles.path_diversity import path_features` instead?

sicihuang replied:
I just saw what you did in path_features_advanced_proposed.py. That would work too!

togawa28 replied:
All your points are right. I'll be clear and consistent in the next revision.

sicihuang commented on mousestyles/path_diversity/path_features_advanced_proposed.py:
```diff
@@ -0,0 +1,71 @@
+from future import print_function, absolute_import, division
+import numpy as np
+import pandas as pd
+from mousestyles.path_diversity.path_features import angle_between
```
Didn't seem like the angle_between funciton was used in this file. Is it necessary to have it here?

togawa28 replied:
I originally had this line to calculate center angles in the path, but since @alannaiverson didn't pick center angles in this file, right now it is unnecessary. If we decide to include center angles in the output of this function, we need it.

alannaiverson replied:
Yeah most of these imports were just me testing input at a different place in my ipynb. This wasn't meant to be a final version of anything, it's an exploration file which is why the imports are different.

sicihuang commented on mousestyles/path_diversity/tests/test_path_features_advanced_proposed.py:
```diff
@@ -0,0 +1,39 @@
+    like_path3 = pd.DataFrame({'t': [2], 'x': [5], 'y': [3], 'isHB': [True]})
+
+    with pytest.raises(TypeError) as excinfo:
+        path_features.compute_angles(like_path, True)
```
Probably should do `from mousestyles.path_diversity import path_features` at the top.

sicihuang commented on mousestyles/path_diversity/tests/test_path_features_advanced_proposed.py:
```diff
@@ -0,0 +1,39 @@
+
+from mousestyles import data
+from mousestyles.path_diversity import path_index
+from mousestyles.path_diversity import filter_path
```
Was any function from path_index or filter_path used in this file?

togawa28 replied:
No. We should remove those.

togawa28 commented:
Guys, I updated path_features_advanced.py and newly added tests/test_path_features_advanced.py, based on our discussions above. I intentionally didn't make any changes in @alannaiverson's file for comparison purpose. Main updates are:
1. main functions returns all features (i.e. includes center angles, area covered by path.)
2. radius and center angles are returned as a list (no data reduction): we could take sum or average outside of the function.
3. modified tests for outputs.
Your any comments (especially on the updates above & styles issue) will be appreciated.

alannaiverson commented on mousestyles/path_diversity/path_features_advanced.py:
```diff
@@ -0,0 +1,90 @@
+
+    return({'radius': radius, 'center_angles': center_angles,
+            'area_cov': area_cov, 'area_rec': area_rec,
+            'abs_distance': abs_distance})
```
From what I see, the only style issue is there is no blank line at the bottom. If you add a line with no white spaces or characters to the bottom of this and the test file, they should pass make style checks!

togawa28 replied:
It seems that I've already had blank line at the bottom which is not shown here with some reason.

alannaiverson commented on mousestyles/path_diversity/path_features_advanced.py:
```diff
@@ -0,0 +1,90 @@
+        each point in the path. The length equals to the length
+        of the path_obj.
+
+    center_anlges : list
```
angles instead of anlges, just a small typo

jcpalumbo3 commented on mousestyles/path_diversity/tests/test_path_features_advanced.py:
```diff
@@ -0,0 +1,79 @@
+    assert np.abs(np.cos(adv_feats['center_angles'][2])-expected1) < 0.0000001
+    assert np.abs(np.cos(adv_feats['center_angles'][1])-expected2) < 0.0000001
+    assert adv_feats['radius'] == [np.sqrt(3**2+4**2)/2] * 4
+
```
Might need some spaces here to conform to the style above.

togawa28 commented on mousestyles/path_diversity/path_features_advanced.py:
```diff
@@ -0,0 +1,90 @@
+    >>> movement = data.load_movement(1,2,1)
+    >>> adv_features = compute_advanced(movement[5:10])
+    """
+    if not isinstance(path_obj, pd.core.frame.DataFrame):
```
Should be black line here.

togawa28 commented:
@jarrodmillman could you help us to pass Travis CI? Basically our files passed the test in Python 3, but it didn't in python 2. We don't know why and are just wondering if you could find any good recommendation for this.

jarrodmillman commented:
@togawa28 I am looking into it.

jarrodmillman commented:
Rebased and fixed here #203
